File: app/src/main/java/pl/kwako/metroid_map/view/MouseOverMapListener.java
package pl.kwako.metroid_map.view;

import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseWheelEvent;

class MouseOverMapListener extends MouseAdapter implements MouseState {

    private int x;
    private int y;
    private int wheelRotation;
    private final Component component;

    public MouseOverMapListener(Component component) {
        this.component = component;
    }

    @Override
    public void mouseMoved(MouseEvent e) {
        x = e.getX();
        y = e.getY();

        component.revalidate();
        component.repaint();
    }

    @Override
    public void mouseWheelMoved(MouseWheelEvent e) {
        wheelRotation = e.getWheelRotation();

        component.revalidate();
        component.repaint();
    }

    @Override
    public int getX() {
        return x;
    }

    @Override
    public int getY() {
        return y;
    }

    @Override
    public int popWheelRotation() {
        var lastRotation = wheelRotation;

        wheelRotation = 0;

        return lastRotation;
    }
}

File: app/src/main/kotlin/pl/kwako/metroid_map/view/MouseOverMapListener.kt
package pl.kwako.metroid_map.view

import java.awt.Component
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseWheelEvent
import javax.swing.SwingUtilities

class MouseOverMapListener(private val component: Component) : MouseAdapter(), MouseState {

    private var x: Int = 0
    private var y: Int = 0
    private var wheelRotation: Int = 0

    override fun getX(): Int = x

    override fun getY(): Int = y

    override fun popWheelRotation(): Int {
        val lastRotation = wheelRotation
        wheelRotation = 0
        return lastRotation
    }

    override fun mouseMoved(e: MouseEvent) {
        x = e.x
        y = e.y
        SwingUtilities.invokeLater {
            component.revalidate()
            component.repaint()
        }
    }

    override fun mouseWheelMoved(e: MouseWheelEvent) {
        wheelRotation = e.wheelRotation
        SwingUtilities.invokeLater {
            component.revalidate()
            component.repaint()
        }
    }
}
